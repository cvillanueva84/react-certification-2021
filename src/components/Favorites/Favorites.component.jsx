import React from 'react';

import { useFavorites } from '../../utils/hooks/useFavorites';
import Card from '../Card';
import { CardsContainer, Container, Loader } from '../UI';

const Favorites = () => {
  const { favoriteVideos, isLoading } = useFavorites();

  return (
    <Container>
      <CardsContainer subtitle="My favorites">
        {isLoading ? (
          <Loader data-testid="loader" />
        ) : (
          favoriteVideos.map((item) => (
            <Card key={item} item={item} path={`/favorite/${item.id}`} />
          ))
        )}
      </CardsContainer>
    </Container>
  );
};

export default Favorites;
