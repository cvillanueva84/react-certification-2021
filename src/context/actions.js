import Types from '../types';

export const contextValueActions = (dispatch) => ({
  search: (payload) => dispatch({ type: Types.SEARCH, payload }),
  theme: () => dispatch({ type: Types.THEME }),
  auth: (payload) => dispatch({ type: Types.AUTH, payload }),
  authLogout: () => dispatch({ type: Types.LOGOUT }),
  videoList: (payload) => dispatch({ type: Types.VIDEO_LIST, payload }),
});
