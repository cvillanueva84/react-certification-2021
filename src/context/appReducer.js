import { types } from '../types/types';

export const appReducer = (state = {}, action) => {
  switch (action.type) {
    case types.search:
      return {
        ...state,
        ...action.payload,
      };
    case types.showVideoDetails:
      return {
        ...state,
        showVideoDetails: true,
      };
    case types.showVideoList:
      return {
        ...state,
        showVideoDetails: false,
      };
    default:
      return state;
  }
};
