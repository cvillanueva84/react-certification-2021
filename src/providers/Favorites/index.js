export { useFavorites, FavoritesProvider } from './Favorites.provider';
