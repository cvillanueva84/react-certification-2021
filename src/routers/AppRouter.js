import React, { useContext } from 'react';
import { BrowserRouter as Router, Switch } from "react-router-dom";
import { GlobalContext } from '../components/Context';
import { LoginView } from '../components/LoginView';
import { DashboardRoutes } from './DashboardRoutes';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';

export const AppRouter = () => {

    const { myStateReducer } = useContext(GlobalContext);

    return (
        <Router>
            <>
                <Switch>
                    <PublicRoute 
                        exact path="/login" 
                        component={LoginView} 
                        isAuthenticated={myStateReducer.logged}
                    />

                    <PrivateRoute 
                        path="/" 
                        component={DashboardRoutes} 
                        isAuthenticated={myStateReducer.logged}
                    />
                </Switch>
            </>
        </Router>
    )
}
