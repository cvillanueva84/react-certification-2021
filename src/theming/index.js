const lightTheme = {
  body: '#E5E5E5',
  cardBg: '#fff',
  cardBgHover: 'rgb(240, 240, 240)',
  color: '#000',
  headerBg: '#1C5476',
  textTiny: 'rgb(70, 70, 70)',
};

const darkTheme = {
  body: '#1e1e2f',
  cardBg: '#27293d',
  cardBgHover: '#7b1fa2',
  color: '#fff',
  headerBg: '#7b1fa2',
  textTiny: '#fff',
};

export { lightTheme, darkTheme };
