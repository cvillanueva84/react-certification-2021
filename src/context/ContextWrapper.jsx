import React, { useCallback, useReducer } from 'react';
import { AppContext } from './appContext';
import { appReducer } from './appReducer';
import Types from '../types';

const init = {
  search: 'wizeline',
  theme: true,
  isAuthenticated: false,
  auth: {},
  videoList: [],
};

export const ContextWrapper = ({ children }) => {
  const [state, dispatch] = useReducer(appReducer, init);

  const contextValue = {
    state,
    auth: useCallback((payload) => dispatch({ type: Types.AUTH, payload }), []),
    search: useCallback((payload) => dispatch({ type: Types.SEARCH, payload }), []),
    theme: useCallback(() => dispatch({ type: Types.THEME }), []),
    authLogout: useCallback(() => dispatch({ type: Types.LOGOUT }), []),
    videoList: useCallback(
      (payload) => dispatch({ type: Types.VIDEO_LIST, payload }),
      []
    ),
  };

  return <AppContext.Provider value={contextValue}>{children}</AppContext.Provider>;
};
