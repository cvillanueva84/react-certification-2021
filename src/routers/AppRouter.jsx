import React, { useContext, useEffect, useState } from 'react';
import { Switch, Redirect, Route } from 'react-router-dom';

import Favorites from '../components/Favorites';
import FavoriteVideoDetails from '../components/FavoriteVideoDetails';
import Home from '../components/Home/Home.component';
import Login from '../components/Login/Login.component';
import VideoDetails from '../components/VideoDetails';
import { AppContext } from '../context/appContext';
import { firebase } from '../firebase/firebase.config';
import { types } from '../types/types';
import { PrivateRoute } from './PrivateRoute';

export const AppRouter = () => {
  const { dispatch } = useContext(AppContext);
  const [isLoggedIn, setIsLoggedIn] = useState(false);

  useEffect(() => {
    firebase.auth().onAuthStateChanged(async (user) => {
      if (user?.uid) {
        dispatch({
          type: types.auth,
          payload: {
            isAuthenticated: true,
            auth: {
              uid: user.uid,
              email: user.email,
            },
          },
        });
        setIsLoggedIn(true);
      } else {
        setIsLoggedIn(false);
      }
    });
  }, [dispatch]);

  return (
    <Switch>
      <Route path="/" exact component={Home} />
      <Route
        path="/login"
        exact
        component={() => (isLoggedIn ? <Redirect to="/" /> : <Login />)}
      />
      <Route path="/video/:id" exact component={VideoDetails} />
      <PrivateRoute
        isAuthenticated={isLoggedIn}
        component={Favorites}
        path="/favorites"
        exact
      />
      <PrivateRoute
        isAuthenticated={isLoggedIn}
        component={FavoriteVideoDetails}
        path="/favorite/:id"
        exact
      />
      <Redirect to="/" />
    </Switch>
  );
};
