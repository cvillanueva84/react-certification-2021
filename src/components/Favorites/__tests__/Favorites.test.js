import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { HashRouter } from 'react-router-dom';

import Favorites from '../index';
import { AppContext } from '../../../context/appContext';
import { darkTheme, lightTheme } from '../../../theming';
import { GlobalStyles } from '../../../GlobalStyles.styled';
import { AppRouter } from '../../../routers/AppRouter';

describe('Testing Favorites component', () => {
  const initialState = {
    search: 'wizeline',
    theme: true,
    isAuthenticated: false,
    auth: {},
    videoList: [],
  };
  const contextValue = {
    dispatch: jest.fn(),
    state: initialState,
  };
  let tree;

  beforeEach(() => {
    tree = render(
      <AppContext.Provider value={contextValue}>
        <ThemeProvider theme={initialState.theme ? lightTheme : darkTheme}>
          <GlobalStyles />
          <HashRouter>
            <AppRouter />
            <Favorites />
          </HashRouter>
        </ThemeProvider>
      </AppContext.Provider>
    );
  });

  it('should take a snapshop', () => {
    expect(tree).toMatchSnapshot();
  });
});
