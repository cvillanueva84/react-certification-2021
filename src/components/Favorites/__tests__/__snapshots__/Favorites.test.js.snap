// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Favorites component should take a snapshop 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-jSFjdj bNEmGo"
      >
        <div>
          <h2
            center="true"
            class="sc-hBMUJo fdQKmf title py-1"
            weight="300"
          >
            Welcome to the Challenge!
          </h2>
          <div
            class="sc-pNWdM bwqMTH"
          >
            <div
              class="sc-jrsJWt fZykTn"
            >
              <div
                class="sc-gKAaRy lgmLbC"
              >
                <div
                  class="sc-iCoGMd SSGno"
                  data-testid="loader"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-jSFjdj bNEmGo"
      >
        <div>
          <div
            class="sc-pNWdM bwqMTH"
          >
            <h3>
              My favorites
            </h3>
            <div
              class="sc-jrsJWt fZykTn"
            >
              <div
                class="sc-gKAaRy lgmLbC"
              >
                <div
                  class="sc-iCoGMd SSGno"
                  data-testid="loader"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-jSFjdj bNEmGo"
    >
      <div>
        <h2
          center="true"
          class="sc-hBMUJo fdQKmf title py-1"
          weight="300"
        >
          Welcome to the Challenge!
        </h2>
        <div
          class="sc-pNWdM bwqMTH"
        >
          <div
            class="sc-jrsJWt fZykTn"
          >
            <div
              class="sc-gKAaRy lgmLbC"
            >
              <div
                class="sc-iCoGMd SSGno"
                data-testid="loader"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-jSFjdj bNEmGo"
    >
      <div>
        <div
          class="sc-pNWdM bwqMTH"
        >
          <h3>
            My favorites
          </h3>
          <div
            class="sc-jrsJWt fZykTn"
          >
            <div
              class="sc-gKAaRy lgmLbC"
            >
              <div
                class="sc-iCoGMd SSGno"
                data-testid="loader"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
